examples/caa/caa.ipynb

```diff
@@ -0,0 +1,590 @@
+    "    show_progress=True,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "SteeringVector(layer_activations={15: tensor([ 0.0644, -0.0403,  0.0907,  ..., -0.1613, -0.0146,  0.0941],\n",
+      "       dtype=torch.float16)}, layer_type='decoder_block')\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(steering_vector)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's sanity check our vector by evaluating the cosine similarity with the ground truth sycophancy vector."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Cosine similarity: 0.995\n"
+     ]
+    }
+   ],
+   "source": [
+    "original_steering_vector = torch.load(f'vec_layer_15_Llama-2-{model_size}-chat-hf.pt')\n",
+    "our_steering_vector = steering_vector.layer_activations[15]\n",
+    "\n",
+    "def calculate_cosine_similarity(a: torch.Tensor, b: torch.Tensor) -> float:\n",
+    "    \"\"\" Calculate cosine similarity between two vectors \"\"\"\n",
+    "    return torch.sum(a * b) / (a.norm() * b.norm())\n",
+    "\n",
+    "print(f\"Cosine similarity: {calculate_cosine_similarity(original_steering_vector, our_steering_vector):.3f}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Steer with Steering Vectors"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To apply steering vectors to the model, we must have some way of modifying the model's forward pass. The approach taken in the official codebases for Representation Engineering and Contrastive Activation Addition both do this by wrapping the model's blocks. This approach is conceptually simpler but has notable limitations, e.g. being unable to work for other models\n",
+    " \n",
+    "The `steering_vectors` library uses Pytorch hooks instead of model wrappers to modify the underlying model's forward pass. You can read more about hooks at the [official PyTorch documentation](https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_hook.html). "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "+1 steered model: 0.716\n",
+      "Unsteered model: 0.685\n"
+     ]
+    }
+   ],
+   "source": [
+    "hook = steering_vector.patch_activations(\n",
+    "    model,\n",
+    "    multiplier = 1, \n",
+    "    # Steer activations for all tokens\n",
+    "    # NOTE: This was the default behavior in CAA for generating results in the paper, rev 27 Dec 2023\n",
+    "    # Reference: https://github.com/nrimsky/CAA/issues/2\n",
```

Comment:
We can calculate the correct token to apply steering at in this notebook too if it gives similar results. The approach they use in CAA of just finding the last index of `[/INST]` is simple enough to understand IMO, but up to you if you think it's worth it

Reply:
I didn't test comprehensively, but it didn't seem to work well when I tried only patching the generation tokens.

Reply:
It's also a good point that CAA (for whatever reason) uses a slightly different template for generation vs reading. This might be one of the factors leading to the performance gap discussed above. 

I haven't tested this yet but will try to do so soon